components/authz-service/storage/v2/postgres/postgres.go

```diff
@@ -1539,9 +1560,52 @@ func (p *pg) DeleteProject(ctx context.Context, id string) error {
 
 	err = p.singleRowResultOrNotFoundErr(res)
 	if err != nil {
+		// don't error if the project is already in the graveyard.
+		// this is necessary since the status might not have been reported
+		// to cereal on the first attempt, in which case we want this
+		// function to be idempotent.
+		if err == storage_errors.ErrNotFound {
+			gyRes, _ := tx.ExecContext(ctx, `SELECT id FROM iam_projects_graveyard WHERE id=$1`, id)
+			gyErr := p.singleRowResultOrNotFoundErr(gyRes)
+			// if we don't find the project in the graveyard, return the original error
+			if gyErr != nil {
+				return err
+			}
+			// project found in graveyard
+			// abort transaction and report success to cereal. we don't really care about the rollback
+			// since nothing will have happened at this point in the transaction in this case.
+			// log the error if any for posterity though.
+			err := tx.Rollback()
+			if err != nil {
+				p.logger.Warnf("failed to rollback ProjectDelete when already graveyarded for id %q: %s",
+					id, err.Error())
+			}
+			return nil
+		}
 		return err
 	}
 
+	// insert project into graveyard
+	_, err = tx.ExecContext(ctx,
+		`INSERT INTO iam_projects_graveyard (id) VALUES ($1)`, id)
```

Comment:
so to make sure I understand, we delete the project from the projects table, then insert into the graveyard. then the rest of the domain workflows purge the project from their databases. then we finally remove the project from the graveyard.

i was expecting the project to first be moved to the graveyard ➡️the domain work gets completed ➡️ the project gets officially deleted. but i spose in that case, the project would look like it still existed while it was being purged from the domain services. so this works 👍

Reply:
Yep! While it's in the graveyard a new project can't be created with that ID.